docs/hibernating-clusters.md

```diff
@@ -0,0 +1,121 @@
+### Cluster Hibernation Controller
+The cluster hibernation controller is a new controller that will watch ClusterDeployments and ensure that the target cluster's
+machines reflect the hibernating state specified in the ClusterDeployment spec.
+
+High level flow of controller logic:
+
+![hibernation_controller](hibernation_controller.png)
+
+The controller uses the actuator pattern to power up and power down machine instances as well as determining
+machine state (running or stopped).
+
+#### Selecting Cluster Machines
+
+Option 1 (Preferred):
+The hibernation controller relies on the actuator to select machines used by the cluster.
+The actuator, given a ClusterDeployment's InfraID selects machines using a method appropriate to the cloud provider (tags/name prefix/resource group).
+
+Option 2:
+The hibernation controller uses the machine API on the target cluster to determine which machines belong to the cluster. It then stores the machine IDs
+in the clusterdeployment (or a separate CR), then uses those machine IDs to start the cluster again.
+
+#### Hibernation Actuator
+The hibernation controller relies on an actuator to work with cloud provider machines.
+This is the interface for the actuator:
+
+```go
+type HibernationActuator interface {
+  // StopMachines will start machines belonging to the given ClusterDeployment
+  StopMachines(cd *hivev1.ClusterDeployment) error
+
+  // StartMachines will select machines belonging to the given ClusterDeployment
+  StartMachines(cd *hivev1.ClusterDeployment) error
+
+  // MachinesRunning will return true if the machines associated with the given
+  // ClusterDeployment are in a running state.
+  MachinesRunning(cd *hivev1.ClusterDeployemnt) (bool, error)
+
+  // MachinesStopped will return true if the machines associated with the given
+  MachinesStopped(cd *hivev1.ClusterDeployment) (bool, error)
+}
+```
+
+#### Handling Incompatible OpenShift Versions
+OpenShift versions earlier than 4.4.8 do not support stopping and starting a cluster without additional work
+to restore etcd. In the case that the cluster deployment's `status.clusterVersionStatus.desired.version` is 
+less than 4.4.8, the hibernation controller will add a condition of `HibernationNotSupported` to the cluster
+deployment. When later reconciling cluster deployments with this condition set, the hibernation controller will
+continue to check the version in case the cluster is upgraded and eventually is able to be hibernated.
+
+#### Approving CSRs
+In the case that CSRs must be approved for a cluster that has had its certificates expired while hibernating,
+we should follow similar checks as the [cluster machine approver](https://github.com/openshift/cluster-machine-approver/blob/0f50c7bfe9b309ce01937274598f5a807d9545df/csr_check.go)
+to ensure we are not introducing an additional security exposure.
+
+### Changes to Other Controllers
+
+The following controllers need to be modified to skip processing a cluster in a hibernating state:
```

Comment:
The unreachable condition should already disable all of these (if not that's a bug), I think we can safely just let unreachable do it's job here.

Reply:
:+1:

Reply:
Cool, so no change to existing controllers. Is there anything we should do when the cluster is back up running?

Reply:
Hmm yeah it could take a few hours before unreachable clears, maybe that is a step that needs to happen on resume.

Reply:
The unreachable controller will check on every change to the ClusterDeployment. So if there was a condition change to indicate that the cluster is now running, then the unreachable controller will check soon after that.